go/ct/gen/block_context.go

```diff
@@ -13,43 +13,153 @@
-		randomOffset = rnd.Uint64()
+	var blockNumber uint64
+	if revisionNumberLength != 0 {
+		blockNumberRangeSolver.AddUpperBoundary(revisionNumber + revisionNumberLength)
 	}
-	blockNumber := revisionNumber + randomOffset
 
-	chainId := common.RandU256(rnd)
-	coinbase, err := common.RandAddress(rnd)
+	// if any relevant variable is already bound, then we need to constraint the block number to be generated.
+	for _, offsetConstraint := range b.variablesOffsetConstraints {
+		if assignedValue, isBound := assignment[offsetConstraint.lower.variable]; isBound {
+
+			assignedValueI64 := int64(assignedValue.Uint64())
+			lowerBound := assignedValueI64 - offsetConstraint.upper.offset
+			upperBound := assignedValueI64 + offsetConstraint.lower.offset - 1
+
+			// if the range is only one number ( aka diff 2 ), then the assigned value must be within the range.
+			diff := offsetConstraint.upper.offset - offsetConstraint.lower.offset
+			isFixedValue := diff == 2 || diff == -2
+			isSameAsfixed := offsetConstraint.lower.offset-1 == assignedValueI64
+			if isFixedValue && !isSameAsfixed {
+				return st.BlockContext{}, ErrUnsatisfiable
+			} else if isFixedValue && isSameAsfixed {
+				blockNumberRangeSolver.AddEqualityConstraint(uint64(upperBound - 1))
+			}
+			// in case of desired out of range and bound value too small.
+			if lowerBound < 0 {
+				lowerBound = 0
+			}
+
+			blockNumberRangeSolver.AddLowerBoundary(uint64(lowerBound))
+			blockNumberRangeSolver.AddUpperBoundary(uint64(upperBound))
+
+			// if the range solver is now unsatisfiable, it is most likely because the assigned value is out of the range
+			// of the first number of the current revision - 256.
+			if !blockNumberRangeSolver.IsSatisfiable() {
+				return st.BlockContext{}, ErrUnsatisfiable
+			}
+		}
+	}
+
+	// generate block number
+	blockNumber, err = blockNumberRangeSolver.Generate(rnd)
 	if err != nil {
 		return st.BlockContext{}, err
 	}
+
+	variableRangeSolver := NewRangeSolver(math.MinInt64, int64(blockNumber))
+	baseSolver := NewRangeSolver(math.MinInt64, int64(blockNumber))
+	// for all non bound relevante variables, assign them a value based on the constraints.
+	for i, offsetConstraint := range b.variablesOffsetConstraints {
+		if _, isBound := assignment[offsetConstraint.lower.variable]; !isBound {
+			variableRangeSolver.Restore(baseSolver)
+			if difference := offsetConstraint.lower.offset - offsetConstraint.upper.offset; difference == 2 || difference == -2 {
+				// if the difference between the two offsets is 2, then we can only generate a fix value.
+				variableRangeSolver.AddEqualityConstraint(offsetConstraint.lower.offset - 1)
+			} else if offsetConstraint.lower.offset > 1 && offsetConstraint.upper.offset < 256 {
+				// if lower offset is greater than 1 and upper is less than 256 we generate in range.
+				variableRangeSolver.AddLowerBoundary(offsetConstraint.upper.offset)
+				variableRangeSolver.AddUpperBoundary(offsetConstraint.lower.offset)
+			} else {
```

Comment:
There is something fishy in this code. The member `kind` of the `variableInequality` is never accessed.
I am not sure if this does the right thing for all scenarios, since the value is never used.  It is quite complicated as well. 

I presume this code could be simplified with the help of some member function in the `variableInequality`, and/or in the `constraintPair`, where get the bounded range.
Such a function would be much easier to unit test as well.  And  then the block in the loop turns into:
```
if _, isBound := assignment[offsetConstraint.lower.variable]; !isBound {
     newValue, err := offsetConstraint.Generate(rnd)
```

Reply:
As discussed, here we are dealing with the tradeoff between implementing a generator that can deal with inequalities vs the ad-hoc implementation for this routine. 
because of time constraints, we leave the ad-hoc implementation